### DISCUSSION TITLE
how to custom variogram_function with non-parametric meachine learning models like SVR, RF, MLP etc.

### DISCUSSION THREAD

**Discussion by sunshineYin** [GeoStat-Framework/PyKrige#201] [Q&A]
Hello, 

I want to make modifications on variogram_functions of kriging interpolation. 

Because I think the categories of existing variogram_functions are not really enough, and the fitting effect is also not really good. So I want to use some non-parametric models, such as svr, random forest, mlp, etc. to fit the variogram function, because they have larger capacity.

The question is: in this case, how should I design the variogram function? Because the models are non-parametric here, the input of them are X and y, (that is d and gamma(d)), which are different from existing functions like linear, exponential, etc. Could you give me some help? (maybe some examples)

Thank you a lot!

**Comment by MuellerSeb** [comment]
Hey there,

keep in mind, that variogram functions need to fulfill some mathematically requirements like *conditional negative semidefiniteness* (CNSD), so the estimated kriging variance is always non-negative (see Webster 2007).

For references on valid variogram models see here:
- Hans Wackemagel (2003): Multivariate Geostatistics
- Webster & Oliver (2007): Geostatistics for Environmental Scientists

Also, you can use variogram models from GSTools with PyKrige. That means, you can use all models presented there and the fitting routines provided by GSTools. Have a look here:
- https://geostat-framework.readthedocs.io/projects/gstools/en/stable/examples/05_kriging/02_pykrige_interface.html#sphx-glr-examples-05-kriging-02-pykrige-interface-py
- https://geostat-framework.readthedocs.io/projects/gstools/en/stable/examples/03_variogram/index.html

Beside that, what do you exactly mean with
> Because I think the categories of existing variogram_functions are not really enough, and the fitting effect is also not really good.
> [...] Because the models are non-parametric here [...]

I am a bit puzzled here.

Sebastian

**Reply by sunshineYin** [reply]
Hi Sebastian, 

Thanks for your reply. And you remind me that the estimated kriging variance should be non-negative.

When we use kriging methods for interpolation, we need to fit the variogram, right? And usually, we use some parametric functions like linear, exponential, sphere, etc., and the formula of these functions are certainly pre-difined. But I think these functions may have some shortcomings due to above-mentioned features.

Based on these, I want to use some non-parametric models to fit the variogram, like svr, rf, mlp, etc. But they are non-parametric, and have no certain formula. The input of svr, rf... are d and gamma(d), representing X and y. And the input of exponential, sphere... are some parameters like d, nugget, sill, range... So, they are different, right?

So how to define the non-parametric functions to integrate with pykrige seamlessly?

**Reply by MuellerSeb** [reply]
The interesting question is: How to assure that these "non-parametric functions" are *conditional negative semidefinite*?
![Bildschirmfoto von 2021-07-12 11-53-50](https://user-images.githubusercontent.com/19690642/125268346-e61f0180-e307-11eb-9632-52c638ae4a4d.png)
(from Webster&Oliver 2007, p. 80)

Here gamma is the variogram model and it has to fulfill the condition described above. With a non-parametric model, I don't know how to prove this.

Further: IMHO variogram analyses can be good described with these handful of parameters: nugget, length scale, sill and a shape-parameter. I don't know what to gain if using non-parametric models. Then I would instantly jump to multi-point statistics to gain more information (See Wackernagel 2003).

And: you ***can*** pass a user-defined model, that only takes the spatial distance of two points. This model can be any predefined function (for example your fitted model). There are no sanity checks if the model is suitable.

So if you have generated a function you want to use as your variogram, just pass it:

```python
from pykrige.ok import OrdinaryKriging

def your_model(args=None, r=0):
    # your magic
    # gamma = ...
    return gamma

# your data ...

OK = OrdinaryKriging(*data, variogram_model="custom", variogram_function=your_model)
```

**Reply by sunshineYin** [reply]
Yeah, Sebastian,

 Thank you very much for your patience and reply! 

And I have realized some non-parametric models to fit the variogram. Actually, some models like random forest have good performance, while others not like svr. However, although the fitting effect of random forest is very good, the interpolation result is very poor! And I have no answer now, maybe the non-negative, or CNSD condition?

I may have to take a closer look at its mathematical principles! And thanks again!

**Reply by MuellerSeb** [reply]
I would be really interested in seeing your code or your results. How does your variogram look like?
Would you mind posting some insights here?

Thanks for the good discussion so far.
